openwisp_firmware_upgrader/signals.py

```diff
@@ -0,0 +1,85 @@
+import logging
+
+from django.db import transaction
+
+from .tasks import delete_firmware_files
+
+logger = logging.getLogger(__name__)
+
+
+def _collect_firmware_files(firmware_images):
+    """
+    Collects file paths from firmware images that need to be deleted
+    """
+    paths = []
+    for image in firmware_images:
+        if image.file:
+            paths.append(image.file.name)
+    return paths
+
+
+def delete_build_files(sender, instance, **kwargs):
+    """
+    Delete firmware files when a Build is deleted
+    """
+    logger.info(f"Build delete signal received for build {instance.pk}")
+    firmware_images = instance.firmwareimage_set.all()
+    logger.info(f"Found {firmware_images.count()} firmware images to delete")
```

Comment:
This is making additional query to the database. We can get this value by calling `len()` on `firmware_images`.